src/backend/cuda/kernel/join.hpp

```diff
@@ -73,6 +73,15 @@ namespace cuda
                         blocksPerMatY * X.dims[3],
                         1);
 
+            int maxBlocksY = cuda::getDeviceProp(cuda::getActiveDeviceId()).maxGridSize[1];
+            // Increase the work per block until the condition is satisfied
+            // Increase threads only upto 512
+            while(blocks.y > maxBlocksY) {
+                if(threads.x * threads.y < 512) threads.y *= 2;
```

Comment:
This if statement only ensures that the maximum threads per block is 512. When this becomes false, we simply keep increasing the work per thread to get into limits of max blocks.